lms/lmsweb/views.py

```diff
@@ -81,22 +81,30 @@ def ratelimit_handler(e):
     f'{LIMITS_PER_MINUTE}/minute;{LIMITS_PER_HOUR}/hour',
     deduct_when=lambda response: response.status_code != 200,
 )
-def login():
+def login(login_error: str = None):
+
     if current_user.is_authenticated:
         return get_next_url(request.args.get('next'))
 
     username = request.form.get('username')
     password = request.form.get('password')
     next_page = request.form.get('next')
+    login_error = request.args.get('login_error')
     user = User.get_or_none(username=username)
 
-    if user is not None and user.is_password_valid(password):
-        login_user(user)
-        return get_next_url(next_page)
-    elif user is not None:
-        return redirect(url_for('login', **{'next': next_page}))
-
-    return render_template('login.html')
+    if request.method == 'POST':
+        if user is not None and user.is_password_valid(password):
+            login_user(user)
+            return get_next_url(next_page)
+        elif user is None or user.is_password_valid(password) is False:
+            login_error = 'Invalid username or password'
+            return redirect(
+                url_for(
+                    'login', **{'next': next_page,
+                                'login_error': login_error}),
+            )
+
+    return render_template('login.html', login_error=login_error)
```

Comment:
This code is very sensitive from the security perspective, hence I want it to have test-coverage of 100%. Can you please create login test file that check all the possible scenarios?

@orronai has already written some tests in this regards, maybe you get to the goal of 100% coverage together :)

Reply:
i added tests for login errors and success